File: modules/expo-proximity-sensor/ios/ExpoProximitySensorModule.swift
import ExpoModulesCore
import UIKit

public final class ExpoProximitySensorModule: Module {
  private let eventName = "onProximityChange"
  private var observer: NSObjectProtocol?

  public func definition() -> ModuleDefinition {
    Name("ExpoProximitySensor")

    Events(eventName)

    AsyncFunction("isAvailableAsync") { () -> Bool in
      let device = UIDevice.current
      let wasEnabled = device.isProximityMonitoringEnabled
      device.isProximityMonitoringEnabled = true
      let available = device.isProximityMonitoringEnabled
      if !wasEnabled {
        device.isProximityMonitoringEnabled = false
      }
      return available
    }
    .runOnQueue(.main)

    OnStartObserving(eventName) {
      self.startObserving()
    }

    OnStopObserving(eventName) {
      self.stopObserving()
    }

    OnDestroy {
      self.stopObserving()
    }
  }

  private func startObserving() {
    guard observer == nil else { return }

    let device = UIDevice.current
    device.isProximityMonitoringEnabled = true
    guard device.isProximityMonitoringEnabled else { return }

    sendEvent(eventName, ["near": device.proximityState])
    observer = NotificationCenter.default.addObserver(
      forName: UIDevice.proximityStateDidChangeNotification,
      object: device,
      queue: .main
    ) { [weak self] _ in
      self?.sendEvent(self?.eventName ?? "onProximityChange", ["near": device.proximityState])
    }
  }

  private func stopObserving() {
    if let observer {
      NotificationCenter.default.removeObserver(observer)
      self.observer = nil
    }
    UIDevice.current.isProximityMonitoringEnabled = false
  }
}

File: modules/expo-proximity-sensor/android/src/main/java/expo/modules/proximitysensor/ExpoProximitySensorModule.kt
package expo.modules.proximitysensor

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import androidx.core.os.bundleOf
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoProximitySensorModule : Module(), SensorEventListener {
  private val eventName = "onProximityChange"
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()
  private val sensorManager: SensorManager
    get() = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
  private val proximitySensor: Sensor?
    get() = sensorManager.getDefaultSensor(Sensor.TYPE_PROXIMITY)

  override fun definition() = ModuleDefinition {
    Name("ExpoProximitySensor")

    Events(eventName)

    AsyncFunction("isAvailableAsync") {
      proximitySensor != null
    }

    OnStartObserving(eventName) {
      proximitySensor?.let { sensor ->
        sensorManager.registerListener(this@ExpoProximitySensorModule, sensor, SensorManager.SENSOR_DELAY_NORMAL)
      }
    }

    OnStopObserving(eventName) {
      sensorManager.unregisterListener(this@ExpoProximitySensorModule)
    }

    OnDestroy {
      sensorManager.unregisterListener(this@ExpoProximitySensorModule)
    }
  }

  override fun onSensorChanged(event: SensorEvent) {
    val distance = event.values.firstOrNull() ?: return
    sendEvent(eventName, bundleOf("near" to (distance < event.sensor.maximumRange)))
  }

  override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) = Unit
}
